lib/descriptor/parser.js

```diff
@@ -0,0 +1,575 @@
+/*!
+ * parser.js -  descriptor parser for bcoin
+ * Copyright (c) 2023, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const HD = require('../hd/hd');
+const {HARDENED} = require('../hd/common');
+// const Descriptor = require('./descriptor');
+const KeyRing = require('../primitives/keyring');
+const PKDescriptor = require('./type/pk');
+const PKHDescriptor = require('./type/pkh');
+const WPKHDescriptor = require('./type/wpkh');
+const SHDescriptor = require('./type/sh');
+const WSHDescriptor = require('./type/wsh');
+const ComboDescriptor = require('./type/combo');
+const AddressDescriptor = require('./type/addr');
+const MultisigDescriptor = require('./type/multisig');
+const RawDescriptor = require('./type/raw');
+const Network = require('../protocol/network');
+const Address = require ('../primitives/address');
+const Script = require('../script/script');
+const {isCompressedEncoding} = require('../script/common');
+const {
+  MAX_SCRIPT_PUSH,
+  MAX_MULTISIG_PUBKEYS
+} = require('../protocol/consensus');
+
+const {
+  ConstPubkeyProvider,
+  OriginPubkeyProvider,
+  HDPubkeyProvider
+} = require('./pubkeyprovider');
+
+const {
+  scriptContext,
+  deriveType,
+  descType,
+  strip,
+  giveExpr,
+  isHex,
+  checkChecksum
+} = require('./common');
+
+/**
+ * DescriptorParser
+ * Parser class for descriptor
+ * @property {Descriptor} descriptor
+ * @property {Network} network
+ */
+
+class DescriptorParser {
+  /**
+   * @constructor
+   * @param {Object} options
+   */
+
+  constructor(options) {
+    this.descriptor = null;
+    this.network = null;
+
+    if (options) {
+      this.fromOptions(options);
+    }
+  }
+
+  /**
+   * Inject properties from options object.
+   * @param {Object} options
+   */
+
+  fromOptions(options) {
+    this.descriptor = options.descriptor;
```

Comment:
Do some validation here. Check the type before storing. If it's a string then parse it.